core/src/memory.rs

```diff
@@ -329,7 +437,9 @@ impl AllocationsContext {
             match Interval::<WasmPage>::try_from(start..end) {
                 Ok(interval) if interval.len() >= pages => break,
                 Err(TryFromRangeError::IncorrectRange) => {
-                    return Err(IncorrectAllocationDataError.into())
+                    unreachable!(
+                        "Allocated memory pages or memory size are incorrect, but we checked it in validate memory param, start: {start:?}, end: {end:?}"
+                    )
```

Comment:
This invariant should be upheld by `valdidate_memory_params` upon creation of object.